pkgs/stdenv/linux/default.nix

```diff
@@ -284,21 +282,63 @@ in
     };
   })
 
-
   # 2nd stdenv that contains our own rebuilt binutils and is used for
-  # compiling our own Glibc.
+  # compiling our own gcc.
   #
   # resulting stage2 stdenv:
-  # - coreutils, glibc, gcc: from bootstrapFiles
-  # - binutils: from nixpkgs, built by bootstrapFiles toolchain
+  # - coreutils, glibc: from bootstrapFiles
+  # - binutils, gcc: from nixpkgs, built by bootstrapFiles toolchain
   (prevStage: stageFun prevStage {
     name = "bootstrap-stage2";
 
     overrides = self: super: {
       inherit (prevStage)
         ccWrapperStdenv
-        gcc-unwrapped coreutils gnugrep
-        perl gnum4 bison;
+        bash bison gnum4 coreutils gnugrep
+        patchelf perl texinfo xz;
+
+      ${localSystem.libc} = getLibc prevStage;
+
+      gcc-unwrapped =
+        let makeStaticLibrariesAndMark = pkg:
+              lib.makeOverridable (pkg.override { stdenv = self.makeStaticLibraries self.stdenv; })
+                .overrideAttrs (a: { pname = "${a.pname}-stage2"; });
+        in super.gcc-unwrapped.override {
+        # Link GCC statically against GMP etc.  This makes sense because
+        # these builds of the libraries are only used by GCC, so it
+        # reduces the size of the stdenv closure.
+        gmp = makeStaticLibrariesAndMark super.gmp;
+        mpfr = makeStaticLibrariesAndMark super.mpfr;
+        libmpc = makeStaticLibrariesAndMark super.libmpc;
+        isl = makeStaticLibrariesAndMark super.isl_0_20;
+        # Use a deterministically built compiler
+        # see https://github.com/NixOS/nixpkgs/issues/108475 for context
+        reproducibleBuild = true;
+        profiledCompiler = false;
+      };
```

Comment:
```suggestion
        super.gcc-unwrapped.override {
        # Use a deterministically built compiler
        # see https://github.com/NixOS/nixpkgs/issues/108475 for context
        reproducibleBuild = true;
        profiledCompiler = false;
      };
```

Reply:
There is no need for this hack if we're building gcc twice -- you should remove it from stage3 as well (github won't let me make a ```suggestion for that one).

Reply:
I don't quite understand the reason for building static libraries in the first place, thus cargo-culted it as is in both places. My vague understanding is that we do it to not confuse `nixpkgs` variants of the libraries with `bootstrapTools` variants of the same libraries.

I would like to clean it up separately (probably outside this PR, as it's already quite big) and remove the need for static builds entirely. I think we can do it in a similar fashion we sorted `libstdc++.so` clash in this PR: remove unversioned .so files completely and possibly move out versioned `.so` or `.a` files to a private directory. But I need to make sure those libraries are not used for linking anywhere before gcc.

Reply:
> I don't quite understand the reason for building static libraries in the first place

Yeah I know; it took me a long time to understand it.

I really appreciate the work you did in #208478 to document some of this stuff, but I feel like this PR is very very rushed, without a full understanding of the effects it will have.

> I would like to clean it up separately (probably outside this PR, as it's already quite big)

What's the big hurry?

Reply:
> I really appreciate the work you did in https://github.com/NixOS/nixpkgs/pull/208478 to document some of this stuff, but I feel like this PR is very very rushed, without a full understanding of the effects it will have.

I'm quite confident in final results this PR produces. I'm not confident changing initial gcc build process WRT libraries presented in `bootstrapTools`. I needs a lot of staring at the library search paths in many packages. This PR does not change that. Thus it's a lot easier to reason about the change. Changing it separately is a lot easier for me than glob everything together in a single PR.

> > I would like to clean it up separately (probably outside this PR, as it's already quite big)
> What's the big hurry?

I am in no hurry. We either do it or not.

Reply:
I'm rebuilding #209459 plus `rebootstrap=stdenv.isAarch64` and `gcc=gcc11`.

I believe it should fix the aarch64-gcc10+ problem with zero overhead on non-aarch64 platforms and no technical debt.

It is almost finished on powerpc64le with no problems.  My arm builders are slow so that will take overnight.